### PR TITLE
CADL, convenience method

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
task https://github.com/Azure/autorest.java/issues/1537

Use `GENERATE_CONVENIENCE_METHODS` env variable to enable it for tests. Will switch to CADL side-car when available.

---

sample

```java
    /**
     * The create operation.
     *
     * @param alertConfigName Unique identifier of an alert configuration. This parameter is case-sensitive.
     * @param body Criteria that determine which anomalies should trigger an alert and via which notification
     *     channel(s). This setting can later be applied to a live inference schedule.
     * @throws IllegalArgumentException thrown if parameters fail the validation.
     * @throws com.azure.core.exception.HttpResponseException thrown if the request is rejected by server.
     * @throws ClientAuthenticationException thrown if the request is rejected by server on status code 401.
     * @throws ResourceNotFoundException thrown if the request is rejected by server on status code 404.
     * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
     * @return criteria that determine which anomalies should trigger an alert and via which notification channel(s) on
     *     successful completion of {@link Mono}.
     */
    public Mono<AlertConfig> create(String alertConfigName, AlertConfig body) {
        RequestOptions requestOptions = new RequestOptions();
        return createWithResponse(alertConfigName, BinaryData.fromObject(body), requestOptions)
                .map(Response::getValue)
                .map(r -> r.toObject(AlertConfig.class));
    }

    /**
     * The read operation.
     *
     * @param alertConfigName Unique identifier of an alert configuration. This parameter is case-sensitive.
     * @throws IllegalArgumentException thrown if parameters fail the validation.
     * @throws com.azure.core.exception.HttpResponseException thrown if the request is rejected by server.
     * @throws ClientAuthenticationException thrown if the request is rejected by server on status code 401.
     * @throws ResourceNotFoundException thrown if the request is rejected by server on status code 404.
     * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
     * @return criteria that determine which anomalies should trigger an alert and via which notification channel(s) on
     *     successful completion of {@link Mono}.
     */
    public Mono<AlertConfig> read(String alertConfigName) {
        RequestOptions requestOptions = new RequestOptions();
        return readWithResponse(alertConfigName, requestOptions)
                .map(Response::getValue)
                .map(r -> r.toObject(AlertConfig.class));
    }

    /**
     * The list operation.
     *
     * @param skip The number of records to skip from the list of records based on the sorting field and ordering method
     *     specified. By default, records are ranked by descending created time (UTC).
     * @param maxpagesize The maximum number of records to be returned per page. If more records are requested via the
     *     API, @nextLink will contain the link to the next page.
     * @param sortBy The name of the field on which you want to sort records. By default, records are sorted by created
     *     time (UTC).
     * @param orderBy Determines whether the records will be returned in descending or ascending order. By default,
     *     records are ranked in descending order.
     * @throws IllegalArgumentException thrown if parameters fail the validation.
     * @throws com.azure.core.exception.HttpResponseException thrown if the request is rejected by server.
     * @throws ClientAuthenticationException thrown if the request is rejected by server on status code 401.
     * @throws ResourceNotFoundException thrown if the request is rejected by server on status code 404.
     * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
     * @return the response body on successful completion of {@link Mono}.
     */
    public Mono<PageAlertConfig> list(Long skip, Long maxpagesize, String sortBy, OrderBy orderBy) {
        RequestOptions requestOptions = new RequestOptions();
        requestOptions.addQueryParam("skip", String.valueOf(skip));
        requestOptions.addQueryParam("maxpagesize", String.valueOf(maxpagesize));
        requestOptions.addQueryParam("sortBy", sortBy);
        requestOptions.addQueryParam("orderBy", orderBy.toString());
        return listWithResponse(requestOptions).map(Response::getValue).map(r -> r.toObject(PageAlertConfig.class));
    }
```

### CODE REVIEW

File: cadl-tests/generate.sh [hunk 1]
```diff
@@ -33,6 +33,8 @@ function generate {
 rm -rf src/main/
 rm -rf cadl-output/
 
+export GENERATE_CONVENIENCE_METHODS=true
```
Comment (weidongxu-microsoft) [inline cadl-tests/generate.sh:36]: using it to let tests to always generate convenience methods.

File: javagen/src/main/java/com/azure/autorest/template/ConvenienceAsyncMethodTemplate.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Override
+    protected boolean isConvenienceMethod(ClientMethod method) {
+        return isMethodAsync(method) && isMethodVisible(method);
+    }
+
+    @Override
+    protected boolean isConvenienceMethod(ConvenienceMethod method) {
+        return isMethodAsync(method.getClientMethod()) && isMethodVisible(method.getClientMethod())
+                && method.getClientMethod().getMethodParameters().stream().noneMatch(p -> p.getClientType() == ClassType.Context);
```
Comment (haolingdong-msft) [inline javagen/src/main/java/com/azure/autorest/template/ConvenienceAsyncMethodTemplate.java:34]: Why do we add `Context` for async client method, but not add for sync client method?

Comment (weidongxu-microsoft) [reply javagen/src/main/java/com/azure/autorest/template/ConvenienceAsyncMethodTemplate.java:34]: Async have `Context` in mono/flux.

File: javagen/src/main/java/com/azure/autorest/template/ConvenienceMethodTemplateBase.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+                });
+    }
+
+    /**
+     * Whether the convenience method should be included.
+     *
+     * @param method the convenience method.
+     * @return Whether include the convenience method.
+     */
+    protected abstract boolean isConvenienceMethod(ClientMethod method);
```
Comment (XiaofeiCao) [inline javagen/src/main/java/com/azure/autorest/template/ConvenienceMethodTemplateBase.java:144]: nit, `shouldInclude` would be clearer?

Comment (weidongxu-microsoft) [reply javagen/src/main/java/com/azure/autorest/template/ConvenienceMethodTemplateBase.java]: changed to `isMethodIncluded`.

**Review by haolingdong-msft — APPROVED** [review]

**Pull request merged**